core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -4976,14 +4993,49 @@ private static void computeVersion(ServletContext context) {
      * Version number of this Jenkins.
      */
     public static String VERSION = UNCOMPUTED_VERSION;
-
+    
+    /**
+     * Minimal Java level required to run this Jenkins.
+     * If {@code null}, the version is unknown.
+     */
+    private static @CheckForNull String JAVA_MIN_LEVEL = null;
+    private static VersionNumber DEFAULT_JAVA_MIN_LEVEL = new VersionNumber("8");
```

Comment:
@alexanderrtaylor Jenkins won't start up due to https://github.com/jenkinsci/extras-executable-war/pull/8/ . This change just adds new API which can be used to determine the current requirement. For plugins the behavior is TBD, but it has to be done on the update center side. Once the plugin with custom Java req is installed, you cannot do much excepting failing horribly on startup